File: Tutorials/Codes/Modern_Control_Course/Chapter16/Lesson3/Chapter16_Lesson3.m
% Chapter16_Lesson3.m
%
% Properties of Controllable Canonical Form (CCF) for analysis and design.
%
% This script constructs a SISO controllable canonical form realization,
% computes its controllability matrix, and performs pole placement by
% coefficient matching.
%
% Related MATLAB/Simulink tools:
%   Control System Toolbox: ss, tf, ctrb, place, acker
%   Simulink: State-Space block, Transfer Fcn block

clear; clc;

% D(s) = s^4 + 6 s^3 + 11 s^2 + 6 s + 2
% N(s) = 3 s^2 + 2 s + 1
% a = [a0 a1 ... a_{n-1}], b = [b0 b1 ... b_{n-1}]
a = [2 6 11 6];
b = [1 2 3 0];
n = numel(a);

A = zeros(n,n);
A(1:n-1,2:n) = eye(n-1);
A(n,:) = -a;

B = zeros(n,1);
B(n) = 1;

C = b;
D = 0;

% Manual controllability matrix [B AB ... A^(n-1)B]
Wc = zeros(n,n);
Apow = eye(n);
for k = 1:n
    Wc(:,k) = Apow * B;
    Apow = A * Apow;
end

disp('A ='); disp(A);
disp('B ='); disp(B);
disp('C ='); disp(C);
disp('Wc ='); disp(Wc);
fprintf('rank(Wc) = %d\n', rank(Wc));
fprintf('det(Wc)  = %.0f\n', det(Wc));

% Desired closed-loop poles.
desired_poles = [-2 -3 -4 -5];

% poly returns [1 alpha_{n-1} ... alpha0].
p_des = poly(desired_poles);
alpha = fliplr(p_des(2:end));   % [alpha0 ... alpha_{n-1}]

% CCF coefficient matching for u = -Kx + r:
K = alpha - a;
Acl = A - B*K;

disp('alpha = [alpha0 ... alpha_{n-1}]'); disp(alpha);
disp('K for u = -Kx + r'); disp(K);
disp('eig(A-BK) ='); disp(eig(Acl).');

% Transfer-function check, if Control System Toolbox is available:
% sys = ss(A,B,C,D);
% tf(sys)
% K_place = place(A,B,desired_poles)
% K_acker = acker(A,B,desired_poles)

% Simulink implementation note:
% Use a State-Space block with matrices A, B, C, D above.
% Feed back yK = K*x through a Sum block implementing u = r - K*x.

File: Tutorials/Codes/Modern_Control_Course/Chapter16/Lesson3/Chapter16_Lesson3.py
"""
Chapter16_Lesson3.py

Properties of Controllable Canonical Form (CCF) for analysis and design.

This script constructs a SISO controllable canonical form realization

    D(s) = s^n + a[n-1] s^(n-1) + ... + a[1] s + a[0]
    N(s) = b[n-1] s^(n-1) + ... + b[1] s + b[0]

and demonstrates:
1. companion/CCF matrix construction,
2. controllability matrix determinant and rank,
3. direct pole placement by coefficient matching.

Required library:
    numpy

Optional libraries for larger work:
    scipy.signal, python-control, slycot
"""

import numpy as np


def ccf_matrices(a, b):
    """Return A, B, C, D=0 for CCF using ascending coefficients.

    Parameters
    ----------
    a : array_like, shape (n,)
        Denominator coefficients [a0, a1, ..., a_{n-1}].
    b : array_like, shape (n,)
        Strictly proper numerator coefficients [b0, b1, ..., b_{n-1}].
        Pad with zeros if the numerator degree is lower than n-1.
    """
    a = np.asarray(a, dtype=float).ravel()
    b = np.asarray(b, dtype=float).ravel()
    n = len(a)
    if len(b) != n:
        raise ValueError("b must have the same length as a; pad numerator with zeros.")

    A = np.zeros((n, n), dtype=float)
    A[:-1, 1:] = np.eye(n - 1)
    A[-1, :] = -a

    B = np.zeros((n, 1), dtype=float)
    B[-1, 0] = 1.0

    C = b.reshape(1, -1)
    D = np.array([[0.0]])
    return A, B, C, D


def controllability_matrix(A, B):
    """Return [B, AB, ..., A^(n-1)B]."""
    n = A.shape[0]
    cols = []
    v = B.copy()
    for _ in range(n):
        cols.append(v)
        v = A @ v
    return np.hstack(cols)


def pole_placement_gain_ccf(a, desired_alpha):
    """Coefficient-matching feedback K for u = -K x + r.

    If D_des(s) = s^n + alpha[n-1]s^(n-1)+...+alpha[1]s+alpha[0],
    then K = [alpha0-a0, alpha1-a1, ..., alpha[n-1]-a[n-1]].
    """
    a = np.asarray(a, dtype=float).ravel()
    desired_alpha = np.asarray(desired_alpha, dtype=float).ravel()
    if desired_alpha.shape != a.shape:
        raise ValueError("desired_alpha must have length n.")
    return desired_alpha - a


def polynomial_from_roots_ascending(roots):
    """Return [alpha0, alpha1, ..., alpha_{n-1}] for monic polynomial."""
    descending = np.poly(np.asarray(roots, dtype=float))  # [1, alpha[n-1], ..., alpha0]
    return descending[:0:-1]  # [alpha0, ..., alpha[n-1]]


def main():
    # Example:
    # D(s) = s^4 + 6 s^3 + 11 s^2 + 6 s + 2
    # N(s) = 3 s^2 + 2 s + 1
    a = np.array([2.0, 6.0, 11.0, 6.0])
    b = np.array([1.0, 2.0, 3.0, 0.0])

    A, B, C, D = ccf_matrices(a, b)
    Wc = controllability_matrix(A, B)

    print("A =\n", A)
    print("B =\n", B)
    print("C =\n", C)
    print("Controllability matrix =\n", Wc)
    print("rank(Wc) =", np.linalg.matrix_rank(Wc))
    print("det(Wc) =", round(np.linalg.det(Wc)))

    desired_roots = np.array([-2.0, -3.0, -4.0, -5.0])
    desired_alpha = polynomial_from_roots_ascending(desired_roots)
    K = pole_placement_gain_ccf(a, desired_alpha)

    Acl = A - B @ K.reshape(1, -1)
    print("desired_alpha [alpha0,...,alpha_{n-1}] =", desired_alpha)
    print("K for u = -Kx + r =", K)
    print("closed-loop eigenvalues =", np.sort_complex(np.linalg.eigvals(Acl)))


if __name__ == "__main__":
    main()
